File: src/jyut-dict/logic/sentence/sentenceset.cpp
#include "sentenceset.h"

#include "logic/dictionary/dictionarysource.h"

SentenceSet::SentenceSet(const std::string &source)
    : _source{source}
    , _sourceShortString{DictionarySourceUtils::getSourceShortString(source)}
{
}

SentenceSet::SentenceSet(const std::string &source,
                         const std::vector<Sentence::TargetSentence> &sentences)
    : _source{source}
    , _sourceShortString{DictionarySourceUtils::getSourceShortString(source)}
    , _sentences{sentences}
{
}

std::ostream &operator<<(std::ostream &out, const SentenceSet &sentenceSet)
{
    out << "======== source\n" << sentenceSet.getSource() << "\n";
    for (const auto &sentence : sentenceSet.getSentences()) {
        out << "=========== translation\n" << sentence.sentence << "\n";
        out << "=========== language\n" << sentence.language << "\n";
        out << "=========== directTarget\n" << sentence.directTarget << "\n";
    }
    return out;
}

bool SentenceSet::isEmpty(void) const
{
    return _sentences.empty()
           || std::all_of(_sentences.begin(),
                          _sentences.end(),
                          [](const Sentence::TargetSentence &sentence) {
                              return sentence.sentence == ""
                                     && sentence.language == ""
                                     && sentence.directTarget == false;
                          });
}

bool SentenceSet::pushSentence(const Sentence::TargetSentence &sentence)
{
    _sentences.push_back(sentence);
    return true;
}

const std::string &SentenceSet::getSource(void) const
{
    return _source;
}

const std::string &SentenceSet::getSourceLongString(void) const
{
    return _source;
}

const std::string &SentenceSet::getSourceShortString(void) const
{
    return _sourceShortString;
}

std::span<const Sentence::TargetSentence> SentenceSet::getSentenceSnippet(
    void) const
{
    if (!_snippet.empty()) {
        return _snippet;
    }

    for (int i = 0; i < std::min(static_cast<size_t>(5), _sentences.size());
         ++i) {
        _snippet.emplace_back(_sentences.at(i));
    }

    return _snippet;
}

std::span<const Sentence::TargetSentence> SentenceSet::getSentences() const
{
    return _sentences;
}

File: src/ios/CantoneseDictionary/CantoneseDictionary/logic/sentence/SentenceSet.swift
//
//  SentenceSet.swift
//  CantoneseDictionary
//
//  Created by Aaron on 2026-09-04.
//

struct TargetSentence {
    var sentence: String
    var language: String
    var directTarget: Bool
}

class SentenceSet {
    private var _source: String
    private var _sourceShortString: String
    private var _snippet: [TargetSentence]?
    private var _sentences: [TargetSentence]

    init(source: String, sentences: [TargetSentence]) {
        self._source = source
        // TODO: Actually implement sourceShortString
        self._sourceShortString = String(
            source.split(separator: " ").first ?? ""
        )
        self._sentences = sentences
    }

    func isEmpty() -> Bool {
        _sentences.isEmpty
            || _sentences.allSatisfy({
                $0.sentence.isEmpty && $0.language.isEmpty
                    && $0.directTarget == false
            })
    }

    func appendSentence(_ sentence: TargetSentence) {
        _sentences.append(sentence)
    }

    func getSource() -> String {
        _source
    }

    func getSourceShortString() -> String {
        _sourceShortString
    }

    func getSentenceSnippet() -> [TargetSentence] {
        if let snippet = _snippet {
            return snippet
        }

        _snippet = []
        for i in 0...4 {
            _snippet?.append(_sentences[i])
        }

        return _snippet ?? []
    }

    func getSentences() -> [TargetSentence] {
        _sentences
    }
}
